### PR TITLE
Avoid conditionals inside assertions analyzer

### LINKED ISSUE #2751: Avoid conditionals inside assertions
Category: Design
Default severity: Info
Enabled by default: true

Use cases: The rule should trigger a diagnostic when a conditional is used inside some assertion (e.g. `Assert.AreEqual(a?.B, "ok")`) and should recommend to split the call into 2 assertions (e.g. `Assert.IsNotNull(a)` and then `Assert.AreEqual(a.B, "ok")`).

AB#2051848

### PR DESCRIPTION
**Pull request by fhnaseer**
Fixes #2751
- Analyzers for Assert, CollectionAssert, and StringAssert
- Assert method should avoid conditional access

Example:
// Warning
Assert.AreEqual(a?.b, "something");

// Fix
Assert.IsNotNull(a);
Assert.AreEqual(a.b, "something");

### CODE REVIEW

File: src/Analyzers/MSTest.Analyzers/AnalyzerReleases.Shipped.md [hunk 1]
```diff
@@ -1,4 +1,12 @@
-﻿## Release 3.4.0
+﻿## Release 3.5.0
+
+### New Rules
+
+Rule ID | Category | Severity | Notes
+--------|----------|----------|-------
+MSTEST0026 | Design | Info | AssertionArgsShouldAvoidConditionalAccessRuleId, [Documentation](https://learn.microsoft.com/dotnet/core/testing/mstest-analyzers/mstest0026)
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/AnalyzerReleases.Shipped.md]: This one should go into the unshipped file.

File: test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            public class A
+            {
+                public string? B { get; set; }
+            }
+
+            [TestClass]
+            public class MyTestClass
+            {
+                [TestMethod]
+                public void NonCompliant()
```
Comment (Evangelink) [inline test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs:34]: Please add some compliant tests.

File: test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                {
+                    A? a = new A();
+                    A c = new A();
+                    string? s = "";
+            
+                    [|Assert.AreEqual(s?.Length, 32)|];
```
Comment (Evangelink) [inline test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs:40]: Please add tests where you have a conditional indexer access.

File: src/Analyzers/MSTest.Analyzers/AssertionArgsShouldAvoidConditionalAccessAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+using Microsoft.CodeAnalysis.Operations;
+
+using MSTest.Analyzers.Helpers;
+
+namespace MSTest.Analyzers;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+public sealed class AssertionArgsShouldAvoidConditionalAccessAnalyzer : DiagnosticAnalyzer
+{
+    private static readonly ImmutableArray<string> SupportedMethodNames = ImmutableArray.Create(new[]
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/AssertionArgsShouldAvoidConditionalAccessAnalyzer.cs]: You probably want to check also `AreSame`, `AreNotSame`, all `CollectionAssert` and all `StringAssert`.

File: src/Analyzers/MSTest.Analyzers/Resources.resx [hunk 5]
```diff
@@ -191,6 +191,15 @@
   <data name="AssemblyInitializeShouldBeValidTitle" xml:space="preserve">
     <value>AssemblyInitialize methods should have valid layout</value>
   </data>
+  <data name="AssertionArgsShouldAvoidConditionalAccessDescription" xml:space="preserve">
+    <value>Assert methods should not contain conditional access.</value>
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/Resources.resx]: Please either provide more explanation as to why this change or remove the description

File: test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+                    [|Assert.AreEqual(s?.Length, s?.Length)|];
+                    [|Assert.AreEqual(a?.B?.Length, 32)|];
```
Comment (Evangelink) [inline test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs]: Add some tests using some extra parenthesis (e.g. `Assert.AreEqual(((a?.B)), 32)`), I think the current logic would be missing these.

File: src/Analyzers/MSTest.Analyzers/AnalyzerReleases.Shipped.md (hunk 1 above)
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/AnalyzerReleases.Shipped.md]: Also the category is `Usage` not `Design`

File: test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+using VerifyCS = MSTest.Analyzers.Test.CSharpCodeFixVerifier<
+    MSTest.Analyzers.AssertionArgsShouldAvoidConditionalAccessAnalyzer,
+    Microsoft.CodeAnalysis.Testing.EmptyCodeFixProvider>;
+
+namespace MSTest.Analyzers.UnitTests;
+
+[TestGroup]
+public sealed class AssertionArgsShouldAvoidConditionalAccessAnalyzerTests(ITestExecutionContext testExecutionContext) : TestBase(testExecutionContext)
+{
+    public async Task WhenUsingConditionalsAccess()
```
Comment (Evangelink) [inline test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs]: You forgot to update the baseline.

File: test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs (hunk 3 above)
Comment (fhnaseer) [reply test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs:40]: Added

File: test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs (hunk 6 above)
Comment (fhnaseer) [reply test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs]: Added

File: src/Analyzers/MSTest.Analyzers/AnalyzerReleases.Shipped.md (hunk 1 above)
Comment (fhnaseer) [reply src/Analyzers/MSTest.Analyzers/AnalyzerReleases.Shipped.md]: Updated. In the original issue, it was Design so I used that.

File: test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs (hunk 2 above)
Comment (fhnaseer) [reply test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs:34]: Added

File: src/Analyzers/MSTest.Analyzers/Resources.resx (hunk 5 above)
Comment (fhnaseer) [reply src/Analyzers/MSTest.Analyzers/Resources.resx:195]: Added

File: src/Analyzers/MSTest.Analyzers/AssertionArgsShouldAvoidConditionalAccessAnalyzer.cs (hunk 4 above)
Comment (fhnaseer) [reply src/Analyzers/MSTest.Analyzers/AssertionArgsShouldAvoidConditionalAccessAnalyzer.cs]: Added

File: src/Analyzers/MSTest.Analyzers/Resources.resx (hunk 5 above)
Comment (fhnaseer) [reply src/Analyzers/MSTest.Analyzers/Resources.resx]: removed,

File: test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs (hunk 6 above)
Comment (fhnaseer) [reply test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs]: Added,

File: src/Analyzers/MSTest.Analyzers/AssertionArgsShouldAvoidConditionalAccessAnalyzer.cs [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        });
+    }
+
+    private static void AnalyzeOperation(OperationAnalysisContext context, INamedTypeSymbol assertSymbol)
+    {
+        var invocationOperation = (IInvocationOperation)context.Operation;
+
+        // This is not an invocation of the expected assertion methods.
+        if (!SupportedMethodNames.Contains(invocationOperation.TargetMethod.Name)
+            || !SymbolEqualityComparer.Default.Equals(assertSymbol, invocationOperation.TargetMethod.ContainingType)
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/AssertionArgsShouldAvoidConditionalAccessAnalyzer.cs]: This is a bit fragile as it would allow to mix method name and types (e.g. `CollectionAssert` and `IsTrue`) but I guess that's probably fine for what we do here.

File: src/Analyzers/MSTest.Analyzers/AssertionArgsShouldAvoidConditionalAccessAnalyzer.cs [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+            || !FindConditionalArguments(invocationOperation))
+        {
+            return;
+        }
+
+        context.ReportDiagnostic(invocationOperation.CreateDiagnostic(Rule));
+    }
+
+    private static bool FindConditionalArguments(IInvocationOperation invocationOperation)
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/AssertionArgsShouldAvoidConditionalAccessAnalyzer.cs]: Nit: `HasAnyConditionalAccessOperationChild`

File: src/Analyzers/MSTest.Analyzers/Resources.resx [hunk 10]
```diff
@@ -191,6 +191,12 @@
   <data name="AssemblyInitializeShouldBeValidTitle" xml:space="preserve">
     <value>AssemblyInitialize methods should have valid layout</value>
   </data>
+  <data name="AssertionArgsShouldAvoidConditionalAccessMessageFormat" xml:space="preserve">
+    <value>Assert methods should not contain conditional access. First assert for null and then do the other assertion.</value>
+  </data>
+  <data name="AssertionArgsShouldAvoidConditionalAccessTitle" xml:space="preserve">
+    <value>Assertion arguments should avoid conditional access</value>
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/Resources.resx]: What about `Avoid conditional access in assertions`? cc @nohwnd

File: src/Analyzers/MSTest.Analyzers/Resources.resx [hunk 11]
```diff
@@ -191,6 +191,12 @@
   <data name="AssemblyInitializeShouldBeValidTitle" xml:space="preserve">
     <value>AssemblyInitialize methods should have valid layout</value>
   </data>
+  <data name="AssertionArgsShouldAvoidConditionalAccessMessageFormat" xml:space="preserve">
+    <value>Assert methods should not contain conditional access. First assert for null and then do the other assertion.</value>
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers/Resources.resx]: What about `Prefer asserting nullability as a different assertion`? cc @nohwnd

File: test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+            {
+                public string? S { get; set; }
+                public List<string> T { get; set; } = new();
+            }
+
+            [TestClass]
+            public class MyTestClass
+            {
+                [TestMethod]
+                public void NonCompliant()
```
Comment (Evangelink) [inline test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs:34]: You are missing `AreSame` and `AreNotSame` cases.

File: test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs (hunk 3 above)
Comment (Evangelink) [reply test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs:40]: I meant the case `?[` (see https://learn.microsoft.com/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-)

**Review by Evangelink — COMMENTED** [review]
It'd be good to validate that no diagnostic is raised when using conditional access within `Assert.IsNull` or `Assert.IsNotNull`.

File: test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs (hunk 3 above)
Comment (fhnaseer) [reply test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs:40]: Added,

File: src/Analyzers/MSTest.Analyzers/AssertionArgsShouldAvoidConditionalAccessAnalyzer.cs (hunk 9 above)
Comment (fhnaseer) [reply src/Analyzers/MSTest.Analyzers/AssertionArgsShouldAvoidConditionalAccessAnalyzer.cs]: Updated

File: test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs (hunk 12 above)
Comment (fhnaseer) [reply test/UnitTests/MSTest.Analyzers.UnitTests/AssertionArgsShouldAvoidConditionalAccessAnalyzerTests.cs:34]: Added

File: src/Analyzers/MSTest.Analyzers/AssertionArgsShouldAvoidConditionalAccessAnalyzer.cs (hunk 8 above)
Comment (fhnaseer) [reply src/Analyzers/MSTest.Analyzers/AssertionArgsShouldAvoidConditionalAccessAnalyzer.cs]: Refactored it. Each assertion type checks only for its own methods

File: src/Analyzers/MSTest.Analyzers/Resources.resx (hunk 11 above)
Comment (fhnaseer) [reply src/Analyzers/MSTest.Analyzers/Resources.resx]: Updated

File: src/Analyzers/MSTest.Analyzers/Resources.resx (hunk 10 above)
Comment (fhnaseer) [reply src/Analyzers/MSTest.Analyzers/Resources.resx]: Updated

**Review by Evangelink — DISMISSED** [review]
LGTM thank you for the nice implementation and the fast updates @fhnaseer :shipit: 

Could you please do a follow-up PR to `dotnet/docs` for this rule? Here is an example PR: https://github.com/dotnet/docs/pull/40785

File: src/Analyzers/MSTest.Analyzers/Resources.resx [hunk 13]
```diff
@@ -191,6 +191,12 @@
   <data name="AssemblyInitializeShouldBeValidTitle" xml:space="preserve">
     <value>AssemblyInitialize methods should have valid layout</value>
   </data>
+  <data name="AssertionArgsShouldAvoidConditionalAccessMessageFormat" xml:space="preserve">
+    <value>Prefer asserting nullability as a different assertion</value>
```
Comment (nohwnd) [inline src/Analyzers/MSTest.Analyzers/Resources.resx]:
```suggestion
    <value>Prefer adding an additional assertion that checks for null</value>
```

**Review by Evangelink — APPROVED** [review]
